File: frontend/src/games/orapa-mine/colors.ts
/**
 * Port TypeScript de `colors.py` (mêmes règles et mêmes hypothèses
 * documentées côté Python) : nom de la couleur de sortie du rayon selon
 * l'ensemble des couleurs de gemme touchées.
 */

import { type Color } from "./types";

// Ordre canonique utilisé pour construire une clé de recherche stable
// (Python utilise un frozenset, non ordonné par nature ; ici on trie
// simplement selon cet ordre fixe avant de rechercher dans les tables).
const CANONICAL_ORDER: Color[] = ["RED", "YELLOW", "BLUE", "WHITE"];

const SINGLE_NAME: Record<Color, string> = {
  RED: "rouge",
  YELLOW: "jaune",
  BLUE: "bleu",
  WHITE: "blanc",
};

const PAIR_MIX: Record<string, string> = {
  "RED,WHITE": "rose",
  "YELLOW,WHITE": "jaune clair",
  "BLUE,WHITE": "bleu clair",
  "RED,YELLOW": "orange",
  "YELLOW,BLUE": "vert",
  "RED,BLUE": "violet",
};

const TRIPLE_WITH_WHITE_MIX: Record<string, string> = {
  "RED,BLUE,WHITE": "violet clair",
  "YELLOW,BLUE,WHITE": "vert clair",
  "RED,YELLOW,WHITE": "orange clair",
};

export function resolveRayColorName(touched: ReadonlySet<Color>): string {
  if (touched.size === 0) return "transparent";

  const sorted = CANONICAL_ORDER.filter((c) => touched.has(c));
  if (sorted.length === 1) return SINGLE_NAME[sorted[0]];
  if (sorted.length === 4) return "gris";
  if (sorted.length === 3 && sorted.join(",") === "RED,YELLOW,BLUE") return "noir";

  const key = sorted.join(",");
  if (sorted.length === 2 && key in PAIR_MIX) return PAIR_MIX[key];
  if (sorted.length === 3 && key in TRIPLE_WITH_WHITE_MIX) return TRIPLE_WITH_WHITE_MIX[key];

  // Les 15 combinaisons possibles sont couvertes ci-dessus (voir
  // colors.py) : ce point ne devrait jamais être atteint.
  throw new Error(`Combinaison de couleurs inattendue : ${sorted.join(", ")}`);
}

File: src/amusement/engine/orapa_mine/colors.py
"""Couleurs des gemmes et règles de mélange du rayon (Orapa Mine).

Le rayon est transparent au départ et se teinte en touchant des gemmes.
Le livret officiel confirme explicitement :
- une seule couleur touchée -> le rayon prend cette couleur ;
- une couleur + blanc -> version "claire" (rose, jaune clair, bleu clair) ;
- les 4 couleurs de base -> gris ;
- chaque couleur ne compte qu'une fois, même touchée plusieurs fois.

Les schémas du livret (résultats "orange", "vert", "violet" sur les
exemples d'extension Corps noir) indiquent en plus un mélange classique
type peinture pour les paires sans blanc (rouge+jaune=orange,
jaune+bleu=vert, rouge+bleu=violet), et un exemple de dialogue confirme
qu'ajouter du blanc à un mélange de deux couleurs donne sa version
"claire" (rouge+bleu+blanc -> violet clair ; jaune+bleu+blanc -> vert
clair, vu dans deux schémas). Par cohérence, rouge+jaune+blanc est donc
supposé donner "orange clair" (non confirmé littéralement dans le texte
disponible, mais suit le même motif).

Le mélange des 3 couleurs de base sans blanc (rouge+jaune+bleu) n'est
mentionné nulle part dans le livret : plutôt que d'inventer un nom, on
retombe sur une description explicite. À corriger si le nom officiel est
retrouvé (voir docs/plan.md).
"""

from __future__ import annotations

from enum import Enum, auto


class Color(Enum):
    """Les quatre couleurs de gemme de la version de base."""

    RED = auto()
    YELLOW = auto()
    BLUE = auto()
    WHITE = auto()


_PAIR_MIX: dict[frozenset[Color], str] = {
    # Une couleur + blanc -> version claire (confirmé par le livret).
    frozenset({Color.RED, Color.WHITE}): "rose",
    frozenset({Color.YELLOW, Color.WHITE}): "jaune clair",
    frozenset({Color.BLUE, Color.WHITE}): "bleu clair",
    # Deux couleurs primaires sans blanc -> mélange type peinture (déduit
    # des schémas de l'extension Corps noir).
    frozenset({Color.RED, Color.YELLOW}): "orange",
    frozenset({Color.YELLOW, Color.BLUE}): "vert",
    frozenset({Color.RED, Color.BLUE}): "violet",
}

_TRIPLE_WITH_WHITE_MIX: dict[frozenset[Color], str] = {
    # Mélange de 2 couleurs + blanc -> version claire du mélange (confirmé
    # pour violet clair et vert clair par les schémas et le dialogue
    # d'exemple ; orange clair déduit par le même motif).
    frozenset({Color.RED, Color.BLUE, Color.WHITE}): "violet clair",
    frozenset({Color.YELLOW, Color.BLUE, Color.WHITE}): "vert clair",
    frozenset({Color.RED, Color.YELLOW, Color.WHITE}): "orange clair",
}

_ALL_FOUR = frozenset({Color.RED, Color.YELLOW, Color.BLUE, Color.WHITE})


def resolve_ray_color(touched: frozenset[Color]) -> str:
    """Nom de la couleur de sortie du rayon selon les couleurs traversées.

    `touched` est l'ensemble des couleurs de gemme touchées par le rayon
    (chaque couleur ne compte qu'une fois — règle c. du livret). Un rayon
    qui n'a touché aucune gemme colorée (rien, ou seulement des diamants)
    ressort "transparent".
    """
    if not touched:
        return "transparent"
    if len(touched) == 1:
        (only,) = touched
        return {
            Color.RED: "rouge",
            Color.YELLOW: "jaune",
            Color.BLUE: "bleu",
            Color.WHITE: "blanc",
        }[only]
    if touched == _ALL_FOUR:
        return "gris"
    if len(touched) == 2 and touched in _PAIR_MIX:
        return _PAIR_MIX[touched]
    if len(touched) == 3 and touched in _TRIPLE_WITH_WHITE_MIX:
        return _TRIPLE_WITH_WHITE_MIX[touched]
    # Rouge+jaune+bleu sans blanc : non défini par le livret disponible.
    names = ", ".join(
        {Color.RED: "rouge", Color.YELLOW: "jaune", Color.BLUE: "bleu", Color.WHITE: "blanc"}[c]
        for c in sorted(touched, key=lambda c: c.name)
    )
    return f"mélange non documenté ({names})"
